weave/trace_server/calls_query_builder/optimization_builder.py

```diff
@@ -0,0 +1,572 @@
+    pb: "ParamBuilder",
+    table_alias: str,
+) -> Optional[str]:
+    """Creates a LIKE-optimized condition for in operations."""
+    # Check if the left side is a GetField operation on a JSON field
+    if not isinstance(operation.in_[0], tsi_query.GetFieldOperator):
+        return None
+    # Return if right-side isn't non-empty list
+    if (
+        len(operation.in_) != 2
+        or not isinstance(operation.in_[1], list)
+        or len(operation.in_[1]) == 0
+    ):
+        return None
+
+    from weave.trace_server.calls_query_builder.calls_query_builder import (
+        get_field_by_name,
+    )
+
+    field = get_field_by_name(operation.in_[0].get_field_).field
+    if not _can_optimize_string_field(field):
+        return None
+
+    # Create OR conditions for each value
+    like_conditions: list[str] = []
+
+    for value_operand in operation.in_[1]:
+        if (
+            not isinstance(value_operand, tsi_query.LiteralOperation)
+            or not isinstance(value_operand.literal_, str)
+            or not value_operand.literal_
+        ):
+            return None
+
+        like_pattern = f'%"{value_operand.literal_}"%'
+        like_condition = _create_like_condition(field, like_pattern, pb, table_alias)
+        like_conditions.append(like_condition)
+
+    or_sql = "(" + " OR ".join(like_conditions) + ")"
+    if _field_requires_null_check(field):
+        return f"({or_sql} OR {table_alias}.{field} IS NULL)"
+    return or_sql
+
+
+def _timestamp_to_datetime_str(timestamp: int) -> str:
+    """Converts a timestamp to a datetime string."""
+    return datetime.datetime.fromtimestamp(
+        timestamp, tz=datetime.timezone.utc
+    ).strftime("%Y-%m-%d %H:%M:%S.%f")
+
+
+def _create_datetime_optimization_sql(
+    operation: Union[tsi_query.GtOperation, tsi_query.GteOperation],
+    pb: "ParamBuilder",
+    table_alias: str,
+    op_str: str,
+) -> Optional[str]:
+    """Creates SQL for datetime optimization using indexed sortable_datetime column.
+
+    Applies a buffer to the timestamp to make the filter more permissive:
+    - For normal context: Subtracts buffer from timestamp
+    - For NOT context: Adds buffer to timestamp
+    """
+    field_operand, literal_operand = _extract_field_and_literal(operation)
+    if field_operand is None or literal_operand is None:
+        return None
+
+    field_name = field_operand.get_field_
+    if not _can_optimize_datetime_field(field_name):
+        return None
+
+    literal_value = literal_operand.literal_
+
+    if not literal_value or not isinstance(literal_value, (int, float)):
+        return None
+
+    # convert timestamp to datetime_str
+    timestamp = int(literal_value)
+
+    # Apply buffer in appropriate direction based on context
+    buffer_seconds = int(DATETIME_BUFFER_TIME_SECONDS)
+    if NotContext.is_in_not_context():
+        # For NOT context, add buffer to make filter more permissive
+        timestamp += buffer_seconds
```

Comment:
doesn't this buffer direction depend on the direction of the operation?

Reply:
yeah, we only support gt and gte rn. that could be made direction agnostic easily. If we ever support LT ;)

Reply:
If we add lt / lte (which we absolutely should!) does that break anything? Do we have protections against this?

Reply:
We should add a restriction on the operator itself to ensure that this constraint is asserted

Reply:
Yo we can only get here when we are in a `process_gt` or `process_gte` context. you would have to explicitly add a `process_lt` to support LT

Reply:
in which case you would refactor this optimization function

Reply:
if you add `$lt` support tomorrow everything will still work, you won't get the optimization until you also implement the `process_lt` fn, but it won't break

Reply:
Its also typed, would be hard to accidentally do it.

```python
def _create_datetime_optimization_sql(
    operation: Union[tsi_query.GtOperation, tsi_query.GteOperation],
```